### PR TITLE
[lodash] Better chaining wrappers, overload clean-up

### LINKED ISSUE #14282: Lodash's _.mapKeys(object, ...) fails if object is type Object
- [x] I tried using the latest `lodash/index.d.ts` file in this repo and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] I want to talk about `lodash/index.d.ts`.
  - The authors of that type definition are cc/ @chrootsu, @vvakame 

If I call _mapKeys(object, ...) with object being type Object, TypeScript gives the following error when it tries to compile:
`error TS2345: Argument of type 'Object' is not assignable to parameter of type 'List<{}> | Dictionary<{}>'`

I think this should work judging by Lodash's documention: https://lodash.com/docs/4.17.4#mapKeys

There was a similar problem in the past regarding _.some(), found here: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/8630

Could you please look into this?

Thanks.

### LINKED ISSUE #14337: Lodash chain methods uses wrong generic type
- [ x ] I tried using the `@types/xxxx` package and had problems.

- Authors: @bczengel, @chrootsu, @stepancar

// Type definitions for Lo-Dash 4.14

This interface (and others) uses wrong types. Type `<T>` should not be used in the method name, only in the interface.

```
    interface LoDashExplicitArrayWrapper<T> {
        /**
         * @see _.head
         */
        first<T>(): T;
    }
```

To make this code use correct type:

```
let a = _([1, 2, 3])
	.chain()
	.first();
```

Currently this code makes `a` with type `{}`.

Definition of interface should be as below, to make `a` of type `number`.

```
    interface LoDashExplicitArrayWrapper<T> {
        /**
         * @see _.head
         */
        first(): T;
    }
```

### LINKED ISSUE #14748: Lodash missing _.transform as part of the LoDashExplicitObjectWrapper
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @bczengel @chrootsu @stepancar

The typings for LoDash is missing `transform` in the LoDashExplicitObjectWrapper and LoDashExplicitArrayWrapper interfaces. 

When using the following code:
```
_.chain(someObject)
    .transform(<someParams>)
    .value();
```

... I get the following compiler error:
```
Property 'transform' does not exist on type 'LoDashExplicitObjectWrapper<Object>'
```

As per the Lodash docs, `transform` should be available to be chained.

### LINKED ISSUE #18956: Lodash definitions not working on TS 2.5
When using `filter` function, TS 2.5 is giving an error:

```
Type '(recpt: Recipient) => boolean | undefined' has no properties in common with type 'Partial<Recipient>'
```

code looks like this: 

```ts
const recipients: Comunication.Recipient[] = [
  {
    name: 'Prefeitura',
    type: 'orgao',
    groups: ['chat','contato','sms']
  },
]

// ...

_.filter(recipients, (recpt) => (
  recpt.type === 'orgao' && recpt.groups && recpt.groups.indexOf(req.params['group']) !== -1
))
```

same code is working fine in 2.4

### LINKED ISSUE #19228: lodash forEach types broken in v4.14.74
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [ ] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @aj-r 

@ types/lodash v4.14.74 is causing type errors in a project I am working on. v4.14.73 works as expected. `event` is any, but it believes `filter` is a string.
```
forEach(event.filters, (filter, name) => {
    ...
});
```

### PR DESCRIPTION
**Pull request by aj-r**
Fixes #14337, Fixes #19228, Fixes #14748, Fixes #18956, Fixes #14282

Apologies for the massive PR. The fix for #14337 required me to modify the chaining wrappers for every single function, so I figured I should clean up some of the messy overloads along the way (which also fixed several other issues).

Unfortunately, this will be a breaking change in terms of what type arguments are accepted. So if anyone was explicitly passing type arguments (e.g. `_.chain([]).find<_.LoDashExplicitObjectWrapper<any>>('a')`), their code may be broken. But I think in the long run this change should be more helpful than harmful, as it removes the need to explicitly pass type arguments in almost all cases.

Summary of changes:

- The chain wrapper interfaces have been greatly simplified. Instead of many wrapper types (`LoDashImplicitArrayWrapper`, `LoDashImplicitObjectWrapper`, `LoDashImplicitNumberArrayWrapper`, etc.), there are now only 2 wrapper types: `LoDashImplicitWrapper` and `LoDashExplicitWrapper`.
  - This allows the return types to be more accurate (especially for functions like `head` and `find`).
- Cleaned up messy function overloads (removed unnecessary type arguments)
  - Made `iteratee` arguments more consistent - now they all use one of the new `Iteratee` types, unless they have a good reason not to.
- Replaced `Object` and `Function` types with more appropriate types
- Changed many `Dictionary<T>` inputs to just take plain objects, so an index signature is not required

Please fill out this template:

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://lodash.com/docs/4.17.4
- [x] Increase the version number in the header if appropriate. (not appropriate)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by aj-r** [comment]
Checks are now passing @bczengel @chrootsu @stepancar @ericanderson @Ailrun

**Comment by sandersn** [comment]
This is major change, so I don't want to merge without *some* feedback from other authors? Anybody?

**Comment by aj-r** [comment]
@bczengel @chrootsu @stepancar @ericanderson @Ailrun see previous comment

**Comment by RyanCavanaugh** [comment]
Is the diff to the tests file the minimal diff to get it passing?

**Comment by aj-r** [comment]
@RyanCavanaugh probably not - I removed a lot of explicit type parameters that I thought were unnecessary after this change. Are minimal diffs desired for test files?

**Comment by RyanCavanaugh** [comment]
@aj-r I think a minimal test diff would be great so we can understand at a glance what kind of edits will be needed in user code. As-is it *appears* like this is an enormous breaking change to all consumers (though I'm sure it's not!).

**Comment by aj-r** [comment]
@RyanCavanaugh ok, I'll see if I can clean up the diff.

File: types/lodash/lodash-tests.ts [hunk 1]
```diff
@@ -8518,42 +8694,33 @@ namespace TestMin {
...
     result = _.minBy<number>(array, '');
-    result = _.minBy<{a: number}, number>(array, {a: 42});
+    result2 = _.minBy(array2, {a: 42});
 
     result = _.minBy<number>(list);
     result = _.minBy<number>(list, listIterator);
     result = _.minBy<number>(list, '');
-    result = _.minBy<{a: number}, number>(list, {a: 42});
-
-    result = _.minBy<number>(dictionary);
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts]: lodash doesn't support `minBy` with dictionaries, so I removed these tests.

File: types/lodash/lodash-tests.ts [hunk 2]
```diff
@@ -3781,77 +3940,68 @@ namespace TestFilter {
 
         result = _.filter<TResult>(array, listIterator);
         result = _.filter<TResult>(array, '');
-        result = _.filter<TResult>(array, /./);
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts]: lodash doesn't support filtering by regexp.

File: types/lodash/lodash-tests.ts [hunk 3]
```diff
@@ -6022,83 +6239,74 @@ namespace TestBindKey {
...
 
-        let result: _.LoDashExplicitObjectWrapper<SampleResult>;
+        let result: _.LoDashExplicitWrapper<SampleResult>;
 
-        result = _(object).chain().bindKey<SampleResult>('foo', 42, '');
+        result = _(object).chain().bindKey('foo', 42, '');
     }
 }
 
-const createCallbackObj: { [index: string]: string; } = { name: 'Joe' };
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts]: `createCallback` does not exist in lodash v4

**Comment by aj-r** [comment]
@RyanCavanaugh the lodash-tests diff should be a little more manageable now :)

**Comment by aj-r** [comment]
@RyanCavanaugh I'm not sure what to do about this error, but it seems to be unrelated to lodash?

`ERROR: 6:16  no-declare-current-package  Instead of declaring a module with `declare module "mongoose-simple-random"`, write its contents in directly in "index.d.ts". See: https://github.com/Microsoft/dtslint/blob/master/docs/no-declare-current-package.md`

**Comment by aj-r** [comment]
@RyanCavanaugh @sandersn This PR has been open for over a month now, and there are many people waiting on its fixes. Is it possible for a DefinitelyTyped maintainer to review it, since the other lodash authors seem to be unresponsive?

**Comment by aj-r** [comment]
Poke @RyanCavanaugh @sandersn @bczengel @chrootsu @stepancar @ericanderson @Ailrun

**Comment by andy-ms** [comment]
@aj-r Should be good to merge, just fix the merge conflict. Sorry for the delay.

**Comment by aj-r** [comment]
@andy-ms thanks! Merge conflicts resolved. The build is failing but it seems to be related to sequelize, not lodash.

**Comment by andy-ms** [comment]
Thanks!

**Comment by Thorarin** [comment]
Ouch, why were the generic type parameters for `defaultsDeep` removed?

**Comment by thorn0** [comment]
@Thorarin 

> If a type parameter does not appear in the types of any parameters, you don't really have a generic function, you just have a disguised type assertion.

-- [README, Common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes)

**Comment by alan-agius4** [comment]
After this change I am getting ‘Index signature is missing’ when i do _.flatMap({a:1 b:2});

Is this intended?

**Comment by aj-r** [comment]
@alan-agius4 No, that was not indended. However, this code works for me in TS 2.4 and 2.2. What version are you using, and can you provide more context?

`const res: number[] = _.flatMap({ a: 1, b: 2 });`

**Comment by alan-agius4** [comment]
@aj-r  I am using `TypeScript`: `2.5.5`
```
interface Foo {
	bar: number;
	foo: number;
}

const x: Foo = {
	bar: 1,
	foo: 1
};

const y = _.flatMap(x); // this is an error
```

**Comment by dawnmist** [comment]
I am having issues with the definition for omit, typescript version 2.5.3

Using redux, we have a reducer state definition of: 

```
export type Activity = ActivityDataType & {
  Ref: RecordReferenceNumber;
  RefKey: string;
};
export interface ActivitiesByRef {
  [refKey: string]: Activity;
};
```
Trying to use omit to remove 'deleted' keys in the reducer now gets rejected by the compiler. For the deletion action the reducer returns:
```
activitiesById = (state: ActivitiesByRef = {}, action: Action): ActivitiesByRef => {
  // process actions...
  if (action.type === 'DELETE_ACTIVITIES') {
    const deletedRefKeys: string[] = action.payload.deletedIds;
    return omit(state, deletedRefKeys);
  }
}
```

Result is now the compile error for the use of omit:

> Type 'Partial<ActivitiesByRef>' is not assignable to type 'ActivitiesByRef'.
>   Index signatures are incompatible.
>     Type 'Activity | undefined' is not assignable to type 'Activity'.
>       Type 'undefined' is not assignable to type 'Activity'.
>         Type 'undefined' is not assignable to type 'ActivityDataType'.

Given that the keys are variable anyway, the actual end result for omit in this case is actually still `ActivitiesByRef` rather than `Partial<ActivitiesByRef>`. It's just one with (potentially) a few less keys in the object. I can work around it by recasting the return value from omit to ActivitiesByRef again (so `return omit(state, deletedRefKeys) as ActivitiesByRef;`, but it represents a missed use case in the handling of the return definition. (Which may actually represent a Typescript bug, in that "Partial" for objects with variable keys really should be treated as being equal to those objects).

**Comment by thorn0** [comment]
`Partial<{ [key: string]: SomeType; }>` is effectively the same as `{ [key: string]: SomeType | undefined; }`, hence the type incompatibility.  See also re index signatures and strict null checks: https://github.com/Microsoft/TypeScript/issues/13778

For now we can solve this by adding an overload for `omit`:

```ts
        omit<T>(
            object: Dictionary<T>,
            ...props: Many<string | number>[]
        ): Dictionary<T>;
```

**Comment by ericanderson** [comment]
I’m terribly confused. Why aren’t we using 

type Diff<T extends string, U extends string> = ({[P in T]: P } & {[P in U]: never } & { [x: string]: never })[T];  
type Omit<T, K extends keyof T> = {[P in Diff<keyof T, K>]: T[P]};  

From https://github.com/Microsoft/TypeScript/issues/12215#issuecomment-308052919

**Comment by thorn0** [comment]
At least because `omit` supports not only simple property names, but also paths like `foo[0].bar`.

**Comment by ericanderson** [comment]
Ah. Nm. I didn’t read the above code sample.

**Comment by ericanderson** [comment]
@dawnmist ActivitiesByRef is probably typed wrong. Your definition says that no matter what I pass in as the key I will get back an Activity. In reality, I presume, you may also get an undefined. If you update your index definition to reflect this your compiler error will go away.

**Comment by thorn0** [comment]
In general, everything associated with those deep property paths is really difficult to type in a correct and/or practical way. For instance, I tried to improve types for `_.zipObjectDeep` doing something like this:
```ts
declare function zipObjectDeep<T>(
  paths: Array<string | number | Array<string | number>>,
  values: T[]
): NestedDictionary<T>;

interface NestedDictionary<T> {
  [ key: string ]: T | NestedDictionary<T>;
}
```
But how would we use a value of the `NestedDictionary` type in practice?

```ts
const a = zipObjectDeep(['foo'], [1]);
```

What can be done with `a.foo`? We still have to use a type assertion to do anything meaningful with it . Even type guards don't work with index signatures (https://github.com/Microsoft/TypeScript/issues/17567). If a type assertion is still needed how are these bells and whistles better than simple `any` or `object`?

Then I realized that `T | NestedDictionary<T>` isn't even the correct type for values because the paths can also be like `foo[0]`, i.e. I forgot about arrays. Can we fix it?
```ts
interface NestedDictionary<T> {
  [ key: string ]: T | NestedDictionary<T> | Array<T | NestedDictionary<T>>;
}
```
Right? No, because arrays can contain other arrays (`foo[0][0]`), and it's something that TypeScript simply doesn't support (see https://github.com/Microsoft/TypeScript/issues/13431#issuecomment-315724315).

----

The situation with `_.omit` is similar to that. Actually, `Partial<T>` (as well as `Dictionary<T>`) isn't the correct return type. To be 100% correct, we need to use `PartialDeep`, which even is already present in lodash.d.ts. But how practical will it be to use? It looks to me like a great way to get bogged down in type assertions and null checks. Some trade-offs have to be made.

**Comment by aj-r** [comment]
@dawnmist I think the best solution in your case would be to change your ActivitiesByRef` definition to this:

```
export interface ActivitiesByRef {
  [refKey: string]: Activity | undefined;
};
```

Since it's impossible for every key in `ActivitiesByRef` to have a value, you should really be forcing your code to check that it's not `undefined` before using it.

That said, I think @thorn0 's overload makes a lot of sense, too, and we should add it:

```
        omit<T>(
            object: Dictionary<T>,
            ...paths: Many<string | number>[]
        ): Dictionary<T>;
```

**Comment by dawnmist** [comment]
@ericanderson If you pass in a key that exists in the object, you are guaranteed an Activity in return. It is only when the key itself doesn't exist in the object that you'd get undefined. And in that case, it is undefined because it is not there, not because it is stored as an undefined value. So the object type should be right - but access functions would have the undefined as a possible output for invalid keys.
```
 getActivity(state: ActivitiesByRef, id: string): Activity | undefined;
```

That is a standard Redux pattern - storing a list of items in an object with a unique key to identify each item so you can update/remove/add/retrieve items to the list via the key. At no time should a key that _exists_ in the object point to anything except a list item (in this case, an Activity).

Otherwise, what is the point of a mapped type if the type can never be properly mapped? What you are saying is that a typescript declaration of
```
interface Example {
  [key: string]: number;
}
```
is _never_ a valid definition - despite this being a typical example from the Typescript Handbook.

**Comment by thorn0** [comment]
@dawnmist @ericanderson That's discussed in https://github.com/Microsoft/TypeScript/issues/13778 and https://github.com/Microsoft/TypeScript/issues/11238#issuecomment-250562397. TL;DR: those who really want it should add `| undefined` to their index signatures themselves, but it might create more hassle than value.

**Comment by thorn0** [comment]
@aj-r I've added this overload to #20717, but there is something strange going on with Travis there.

**Comment by aj-r** [comment]
@alan-agius4 For that particular use case, you should use `_.values` instead of `_.flatMap`. However, that does point me at a use case that I think you're getting at:

```
interface Foo {
	bar: number;
	foo: number[];
}

const x: Foo = {
	bar: 1,
	foo: [2, 3],
};

const y = _.flatMap(x); // this is an error
```

Unfortunately, I can't find a good way to handle this that doesn't involve an index signature. The best I can do is this:

```
flatMap(
    collection: object | null | undefined
): any[];
```

 I'll add this overload.

**Comment by ericanderson** [comment]
In my code bases, we set index values to be | undefined because code that access them need to do the check and people may forget otherwise. 

I think here in the non-nested case, the code is actually helping prevent bugs. The returned type is no longer valid even your type says “all keys have a defined value”. 

I’m the case where you might Object.keys, you can always add the exclamation point. 

There has long been a discussion about undefined as a value and undefined as in there is no key. This has come up a lot with things like Partial as optionality implies | undefined.

**Comment by aj-r** [comment]
@ericanderson  I'm a fan of putting `| undefined` on index signatures, too. But at the same time, I don't think we should enforce that style on everyone who uses lodash, especially if it is still up for debate. I think it's best for our code to be compatible with both styles if possible.

**Comment by alexlur** [comment]
By the way, this pull request breaks either `_.isFunction` (line 9619) or TypeScript, because the value will not be recognized as a callable expression.

image

**Comment by thorn0** [comment]
`| Function` should be removed from there. `Function`, `Object`, `Number` and so on are interfaces needed only to define methods of the corresponding prototypes. These types shouldn't be used for anything else.

**Comment by aj-r** [comment]
Agreed. I'll fix this in #20795.

**Pull request merged**